controller/src/test/java/io/pravega/controller/store/task/TaskStoreFactoryForTests.java

```diff
@@ -0,0 +1,141 @@
+    }
+
+    /**
+     * Extends {@link ZKTaskMetadataStore} with the calls in {@link TaskMetadataStoreExtensions}.
+     */
+    public static class ZKTaskMetadataStoreForTests extends ZKTaskMetadataStore implements TaskMetadataStoreExtensions {
+
+        private final AtomicReference<CompletableFuture<Void>> latch = new AtomicReference<>(new CompletableFuture<Void>());
+
+        ZKTaskMetadataStoreForTests(CuratorFramework client, ScheduledExecutorService executor) {
+            super(client, executor);
+            this.latch.get().complete(null);
+        }
+
+        @Override
+        public void blockCriticalSection() {
+            latch.set(new CompletableFuture<>());
+        }
+
+        @Override
+        public void unblockCriticalSection() {
+            latch.get().complete(null);
+        }
+
+        @Override
+        public CompletableFuture<Void> lock(final Resource resource,
+                                            final TaskData taskData,
+                                            final String owner,
+                                            final String tag,
+                                            final String oldOwner,
+                                            final String oldTag) {
+            CompletableFuture<Void> future = super.lock(resource, taskData, owner, tag, oldOwner, oldTag);
+            try {
+                this.latch.get().get();
+            } catch (Throwable e) {
+                throw new CompletionException(e);
+            }
+
+            return future;
+        }
+    }
+
+    /**
+     * Extends {@link InMemoryTaskMetadataStore} with the calls in {@link TaskMetadataStoreExtensions}.
+     */
+    public static class InMemoryTaskMetadataStoreForTests extends InMemoryTaskMetadataStore
+            implements TaskMetadataStoreExtensions {
+
+        private final AtomicReference<CompletableFuture<Void>> latch =
+                new AtomicReference<>(new CompletableFuture<Void>());
+
+        InMemoryTaskMetadataStoreForTests(ScheduledExecutorService executor) {
+            super(executor);
+            this.latch.get().complete(null);
+        }
+
+        @Override
+        public void blockCriticalSection() {
+            latch.set(new CompletableFuture<>());
+        }
+
+        @Override
+        public void unblockCriticalSection() {
+            latch.get().complete(null);
+        }
+
+        @Override
+        public CompletableFuture<Void> lock(final Resource resource,
+                                            final TaskData taskData,
+                                            final String owner,
+                                            final String tag,
+                                            final String oldOwner,
+                                            final String oldTag) {
+            CompletableFuture<Void> future = super.lock(resource, taskData, owner, tag, oldOwner, oldTag);
+            try {
+                this.latch.get().get();
```

Comment:
you can do a `join` on the future so that you dont have to surround it with a try catch block